kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+  if self.string_capacity then
+    -- If string_capacity is set, reduce the currently queued byte count by the
+    self.bytes_queued = self.bytes_queued - #self.entries[self.front]
+  end
+  self.entries[self.front] = nil
+  self.front = self.front + 1
+  if self.front == self.back then
+    self.front = 1
+    self.back = 1
+  end
+end
+
+-- Process one batch of entries from the queue.  Returns truthy if entries were processed, falsy if there was an
+-- error or no items were on the queue to be processed.
+function Queue:process_once()
+  local ok, err = self.semaphore:wait(0)
+  if not ok then
+    if err ~= "timeout" then
+      -- We can't do anything meaningful to recover here, so we just log the error.
+      self:log(ERR, 'error waiting for semaphore: %s', err)
+    end
+    return
+  end
+  local data_started = now()
+
+  local entry_count = 1
+
+  -- We've got our first entry from the queue.  Collect more entries until max_delay expires or we've collected
+  -- batch_max_size entries to send
+  while entry_count < self.batch_max_size and (now() - data_started) < self.max_delay and not ngx.worker.exiting() do
+    ok, err = self.semaphore:wait(((data_started + self.max_delay) - now()) / 1000)
+    if not ok and err == "timeout" then
+      break
+    elseif ok then
+      entry_count = entry_count + 1
+    else
+      self:log(ERR, "could not wait for semaphore: %s", err)
+      break
+    end
+  end
+
+  local start_time = now()
+  local retry_count = 0
+  while true do
+    self:log(DEBUG, "passing %d entries to handler", entry_count)
+    ok, err = self.handler(self.handler_conf, {unpack(self.entries, self.front, self.front + entry_count - 1)})
+    if ok then
+      self:log(DEBUG, "handler processed %d entries sucessfully", entry_count)
+      break
+    end
+
+    if not err then
+      self:log(ERR, "handler returned falsy value but no error information")
+    end
+
+    if (now() - start_time) > self.max_retry_time then
+      self:log(
+        ERR,
+        "could not send entries, giving up after %d retries.  %d queue entries were lost",
+        retry_count, entry_count)
+      break
+    end
+
+    self:log(WARN, "handler could not process entries: %s", tostring(err))
+    retry_count = retry_count + 1
+    -- Before trying to send the same batch again, wait for a little while.  The wait time is calculated by taking
+    -- the square of the retry count multiplied by 10 milliseconds, creating an exponential increase of the wait
+    -- time (i.e. 10, 20, 40, 80 ... milliseconds).  The maximum time between retries is capped by the configuration
+    -- parameter max_retry_delay
+    ngx.sleep(math.min(self.max_retry_delay, (retry_count * retry_count) * 0.01))
+  end
+
+  -- Guard against queue shrinkage during handler invocation by using match.min below.
+  for _ = 1, math.min(entry_count, self:count()) do
+    self:delete_frontmost_entry()
+  end
+  if self.queue_full then
+    self:log(INFO, 'queue resumed processing')
```

Comment:
I may have had too much coffee and my brain is acting up but I don't understand what does "resuming processing" is supposed to mean.
Isn't the queue being processed at whatever rate it is supposed to be and when full, the queue will drop older items?

Reply:
Once the queue has run out of capacity, it is dropping entries, but we don't want to log a message for each entry dropped.  Thus, we remember that we logged the "queue out of capacity" message and once entries are removed from the queue by the consuming end, we switch that flag off again.  While this mechanism is not perfect, it prevents the log from being flooded with gazillions of messages.